blockchain/validate.go

```diff
@@ -2165,44 +2259,106 @@ func isTicketCommitP2SH(script []byte) bool {
 	return script[commitAmountEndIdx-1]&0x80 != 0
 }
 
-// calcTicketReturnAmount calculates the required amount to return from a ticket
-// for a given original contribution amount, the price the ticket was purchased
-// for, the vote subsidy (if any) to include, and the sum of all contributions
-// used to purchase the ticket.
+// calcTicketReturnAmounts calculates the required amounts to return from a
+// ticket for the given original contribution amounts, the price the ticket was
+// purchased for, and the vote subsidy (if any) to include.
 //
 // Since multiple inputs can be used to purchase a ticket, each one contributes
 // a portion of the overall ticket purchase, including the transaction fee.
 // Thus, when claiming the ticket, either due to it being selected to vote, or
 // being revoked, each output must receive the same proportion of the total
 // amount returned.
 //
+// After the original contribution amounts are evenly distributed to each
+// output, there may be a remainder left over of 1 to numOutputs - 1 atoms.  For
+// votes, this remainder ends up as part of the transaction fee.  If the
+// automatic ticket revocations agenda is NOT active, then this is the same for
+// revocations and the remainder ends up as part of the transaction fee.  If the
+// automatic ticket revocations agenda IS active, then for revocations each atom
+// in the remainder is added to an output index that is selected in a uniformly
+// random manner, where Hash256PRNG seeded with the previous header bytes is
+// used for the deterministic pseudorandom selection.
+//
 // The vote subsidy must be 0 for revocations since, unlike votes, they do not
 // produce any additional subsidy.
-func calcTicketReturnAmount(contribAmount, ticketPurchaseAmount, voteSubsidy int64, contributionSumBig *big.Int) int64 {
-	// This is effectively equivalent to:
-	//
-	//                  total output amount
-	// return amount =  ------------------- * contribution amount
-	//                  total contributions
-	//
-	// However, in order to avoid floating point math, it uses 64.32 fixed point
-	// integer division to perform:
+func calcTicketReturnAmounts(ticketOuts []*stake.MinimalOutput,
+	ticketPurchaseAmount, voteSubsidy int64, prevHeaderBytes []byte, isVote,
```

Comment:
At first it drew my attention that we'd have to depend on the previous block header here, but since this check _necessarily_ needs to be (at least) a positional one due to requiring the correct `voteSubsidy` for the block that includes the redeemer being checked, this ends up not really being an _additional_ requirement (the only change necessary as implemented here was to pass the parent header bytes).

All in all, a clever trick. :P

Reply:
This is a good call-out.  It does present a situation that is not ideal for the case where we would broadcast version `2` revocations for legacy missed/expired tickets (e.g., tickets that were missed or expired before the agenda activates).  In that case, if those transactions are not accepted into the following block, and remain in the mempool, they will be invalid in any subsequent blocks.  This wouldn't cause an issue with template generation since [they will be invalid per `CheckTransactionInputs` and excluded from the template](https://github.com/decred/dcrd/blob/1210cc2c4eb263409c3b8db8b913a18e1ccf3d33/internal/mining/mining.go#L1774-L1783), but they should probably be evicted from the mempool.  I will make an update for that.

Reply:
I [added logic](https://github.com/decred/dcrd/compare/d96dcf7b6a04e84f615a0bc6f26fde59fbd055f7..202dea7a404321d74d856b23ed6fcc25d22b7f2a) to `pruneStakeTx` to remove revocations from the mempool when a new block is processed if the auto revocations agenda is active to handle this case.